pkg/oomstore/cache.go

```diff
@@ -0,0 +1,175 @@
+package oomstore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/oom-ai/oomstore/internal/database/dbutil"
+	"github.com/oom-ai/oomstore/internal/database/metadata"
+	"github.com/oom-ai/oomstore/internal/database/offline"
+	"github.com/oom-ai/oomstore/pkg/errdefs"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types"
+)
+
+const (
+	Day       = 24 * time.Hour
+	Capacity  = 1000
+	Period    = 5 * time.Minute
+	MinPeriod = 2 * time.Minute
+)
+
+type BufferGroup struct {
+	groupID  int
+	records  []types.StreamRecord
+	lastPush time.Time
+	mu       sync.Mutex
+}
+
+type Cache struct {
+	capacity  int
+	minPeriod time.Duration
+	mu        sync.Mutex
+	quit      chan bool
+
+	ticker *time.Ticker
+	ch     chan types.StreamRecord // buffer channel
+	buffer map[int]*BufferGroup    // group_id -> buffered rows
+}
+
+func (s *OomStore) InitCache(ctx context.Context) {
+	cache := &Cache{
+		// TODO: make Capacity, Period, MinPeriod configurable
+		capacity:  Capacity,
+		minPeriod: MinPeriod,
+		quit:      make(chan bool),
+
+		ch:     make(chan types.StreamRecord),
+		buffer: make(map[int]*BufferGroup),
+		ticker: time.NewTicker(Period),
+	}
+	s.cache = cache
+
+	go func() {
+		for {
+			select {
+			case <-cache.quit:
+				return
```

Comment:
Before exiting, need to check if there are any records that have not been sent.

Reply:
fixed